### PR TITLE
Metric: P&R

### LINKED ISSUE #240: Metric: Precision-Recall, Density-Coverage
**Is your feature request related to a problem? Please describe.**
Hi, I saw you have some common feature-based metrics working nicely on GPU. I was wondering if you would be interested in including a couple metrics that don't comprise the whole image assessment into a single scalar, like FID or KID do. But rather having separated quality and diversity scores to distinguish different failure cases, i.e. realistic but low diversity, low-quality but high diversity.

This metrics are Improved Precision and Recall [(P&R)](https://arxiv.org/abs/1904.06991) and Density and Coverage [(D&C)](https://arxiv.org/abs/2002.09797)

**Describe the solution you'd like**
I found released numpy/sklearn [code](https://github.com/clovaai/generative-evaluation-prdc/blob/master/prdc/prdc.py) that could be kinda straight-away ported to pytorch.

Cheers.

### PR DESCRIPTION
**Pull request by hecoding**
Closes #240 (1/2)  
Adaptation of [released code](https://github.com/clovaai/generative-evaluation-prdc/blob/master/prdc/prdc.py).
Same behaviour as other feature-based metrics (FID, KID, ...)

## Proposed Changes

  - Metric [Improved Precision and Recall (NVIDIA)](https://arxiv.org/abs/1904.06991)
  - Tests
  - Updated README

### CODE REVIEW

File: piq/pr.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+"""
+PyTorch implementation of Improved Precision and Recall (P&R)
+Reference:
+    Kynkäänniemi T. et al. Improved Precision and Recall Metric for Assessing Generative Models,
+    https://arxiv.org/abs/1904.06991
+Credits:
+    https://github.com/clovaai/generative-evaluation-prdc/blob/master/prdc/prdc.py
+"""
+
+from typing import Tuple
```
Comment (zakajd) [inline piq/pr.py:10]:
```suggestion
from typing import Tuple, Optional
```

File: piq/pr.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import torch
+
+from piq.base import BaseFeatureMetric
+from piq.utils import _validate_input
+
+
+def _compute_pairwise_distance(data_x: torch.Tensor, data_y: torch.Tensor = None) -> torch.Tensor:
```
Comment (zakajd) [inline piq/pr.py:17]:
```suggestion
def _compute_pairwise_distance(data_x: torch.Tensor, data_y: Optional[torch.Tensor] = None) -> torch.Tensor:
```

File: tests/test_pr.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
+
+
+def test_forward(features_y_normal, features_x_normal, ) -> None:
+    try:
+        metric = PR()
+        metric(features_y_normal, features_x_normal)
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
```
Comment (zakajd) [inline tests/test_pr.py:40]: No need to catch exceptions, pytest will do it.
We have similar code in other metrics, but will fix this eventually. 
```suggestion
def test_forward(features_y_normal, features_x_normal, ) -> None:
        metric = PR()
        metric(features_y_normal, features_x_normal)
```

File: tests/test_pr.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+@pytest.fixture(scope='module')
+def features_x_constant() -> torch.Tensor:
+    return torch.ones(1000, 20)
+
+
+def test_initialization() -> None:
+    try:
+        PR()
+    except Exception as e:
+        pytest.fail(f"Unexpected error occurred: {e}")
```
Comment (zakajd) [inline tests/test_pr.py:32]: I'll vote to drop this test, because metric init with default parameters is checked in all other test

**Review by zakajd — CHANGES_REQUESTED** [review]
Nice! 
Just a few changes and it's ready to go.

File: piq/pr.py (hunk 2 above)
Comment (hecoding) [reply piq/pr.py]: oops, thanks

File: tests/test_pr.py (hunk 3 above)
Comment (hecoding) [reply tests/test_pr.py:40]: You're right, I just copied KID tests without much thought as they'd technically be the same

File: tests/test_pr.py (hunk 4 above)
Comment (hecoding) [reply tests/test_pr.py:32]: True that.

**Comment by hecoding** [comment]
Cool, thanks for the review @zakajd
Changes are all committed.

File: docs/source/modules.rst [hunk 5]
```diff
@@ -113,6 +113,10 @@ Multi-Scale Intrinsic Distance (MSID)
 .. autoclass:: piq.MSID
     :members:
 
+Improved Precision and Recall (P&R)
+'''''''''''''''''''''''''''''''
```
Comment (denproc) [inline docs/source/modules.rst:117]: The formatting style for rst files requires the title and symbols below be of the same length.
```suggestion
Improved Precision and Recall (P&R)
'''''''''''''''''''''''''''''''''''
```

File: piq/pr.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    Returns:
+        precision: Scalar value of the precision of image sets features.
+        recall: Scalar value of the recall of image sets features.
+
+    References:
+        .. [1] Kynkäänniemi T. et al. (2019).
+           Improved Precision and Recall Metric for Assessing Generative Models.
+           Advances in Neural Information Processing Systems,
+           https://arxiv.org/abs/1904.06991
```
Comment (denproc) [inline piq/pr.py:77]: We unify the documentation style in #243. Please, check out the formatting and change it accordingly (Returns, References) and add Example section.

**Review by denproc — CHANGES_REQUESTED** [review]
Great work!
There are some comments regarding the documentation part + flake8 code style (see the CI test) to be resolved.

File: piq/pr.py (hunk 6 above)
Comment (hecoding) [reply piq/pr.py]: I understand, I must have copied the wrong class for the format. Hope it all abides by the style now.

What do you mean by Example section - add an example not in the docstrings but the README? Wouldn't it be too verbose for a Readme to enumerate all metrics that behave the same way?

File: piq/pr.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    Returns:
+        precision: Scalar value of the precision of image sets features.
+        recall: Scalar value of the recall of image sets features.
+
+    References:
+        Kynkäänniemi T. et al. (2019).
+        Improved Precision and Recall Metric for Assessing Generative Models.
+        Advances in Neural Information Processing Systems,
+        https://arxiv.org/abs/1904.06991
```
Comment (denproc) [inline piq/pr.py:72]: Changes:
- Args of the class are the arguments of the `__init__`. 
- Class is an object and do not have returns, but its functions have. 
- Added Example of usage

```suggestion
    Args:
        nearest_k: Nearest neighbor to compute the non-parametric representation. Shape :math:`1`
        
    Examples:
        >>> loss = PR()
        >>> x = torch.rand(3, 3, 256, 256, requires_grad=True)
        >>> y = torch.rand(3, 3, 256, 256)
        >>> precision, recall = loss(x, y)

    References:
        Kynkäänniemi T. et al. (2019).
        Improved Precision and Recall Metric for Assessing Generative Models.
        Advances in Neural Information Processing Systems,
        https://arxiv.org/abs/1904.06991
```

File: piq/pr.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""Creates non-parametric representations of the manifolds of real and generated data and computes
+        the precision and recall between them.
+
+        Args:
+            real_features: Samples from data distribution. Shape :math:`(N_x, D)`
+            fake_features: Samples from fake distribution. Shape :math:`(N_x, D)`
+        Returns:
+            precision: Scalar value of the precision of the generated images.
+            recall: Scalar value of the recall of the generated images.
```
Comment (denproc) [inline piq/pr.py:94]: The documentation parse the string before semicolon as a Return Type. I would suggest following.
```suggestion
            Scalar value of the precision of the generated images.
            
            Scalar value of the recall of the generated images.
```

**Review by denproc — CHANGES_REQUESTED** [review]
Thanks for the changes. I meant the "Examples" section inside the docstring of the class, see the proposed changes. In this case, the documentation of each class will contain the usage example.

Fyi, you can check the look of the documentation for the class by pressing on the Details link for the readthedocs CI testing.

File: piq/pr.py (hunk 6 above)
Comment (denproc) [reply piq/pr.py]: No worries, previous style for class docstrings was not unified, which could induce some misunderstanding.

**Review by zakajd — APPROVED** [review]

**Review by denproc — APPROVED** [review]
LGTM
